### PR TITLE
feature(stdlib): add snappy codec

### PR DESCRIPTION
**Pull request by MichaHoffmann**
* add 'encode_snappy' and 'decode_snappy'

I have to deal with snappy encoded messages and this function would help me a lot!

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -49,7 +49,14 @@ lua = ["dep:mlua"]
 test = ["string_path"]
 
 # All stdlib functions
-stdlib = ["compiler", "core", "datadog", "dep:aes", "dep:chacha20poly1305", "dep:crypto_secretbox", "dep:ctr", "dep:cbc", "dep:cfb-mode", "dep:ofb", "dep:base16", "dep:nom", "dep:strip-ansi-escapes", "dep:utf8-width", "dep:hex", "dep:seahash", "dep:syslog_loose", "dep:hostname", "dep:zstd", "dep:quoted_printable", "dep:once_cell", "dep:base64", "dep:uuid", "dep:percent-encoding", "dep:uaparser", "dep:rust_decimal", "dep:indexmap", "dep:flate2", "dep:charset", "dep:data-encoding", "dep:hmac", "dep:sha-1", "dep:cidr-utils", "dep:sha-2", "dep:md-5", "dep:url", "dep:woothee", "dep:csv", "dep:roxmltree", "dep:rand", "dep:dns-lookup", "dep:sha-3", "dep:grok", "dep:community-id"]
+stdlib = ["compiler", "core", "datadog", "dep:aes", "dep:chacha20poly1305", "dep:crypto_secretbox",
```
Comment (pront) [inline Cargo.toml:52]: Can you preserve the format here?

**Review by pront — COMMENTED** [review]
Hi @MichaHoffmann,

Thank you for your contribution!

This looks good overall. Note that if there are any bugs with this function, they will likely not be prioritized over other work.

A few things before we merge:
* Please add a CHANGELOG.md entry
* Please make a new PR [here](https://github.com/vectordotdev/vector/tree/master/website/cue/reference/remap/functions) to document these two new fucntions.

File: Cargo.toml (hunk 1 above)
Comment (MichaHoffmann) [reply Cargo.toml:52]: Yeah! I think my language server autoformatted here accidentally!

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,21 +10,22 @@
...
-- `parse_aws_vpc_flow_log` now handles account-id value as a string, avoiding loss of leading zeros and case where value is `unknown` (https://github.com/vectordotdev/vrl/issues/263) 
+- `parse_aws_vpc_flow_log` now handles account-id value as a string, avoiding loss of leading zeros and case where value is `unknown` (https://github.com/vectordotdev/vrl/issues/263)
 
 #### Features
 - `parse_key_value` can now parse values enclosed in single quote characters (https://github.com/vectordotdev/vrl/pull/382)
 - added `pretty` parameter for `encode_json` vrl function to produce pretty-printed JSON string (https://github.com/vectordotdev/vrl/pull/370)
 - added `community_id` function for generation of [V1 Community IDs](https://github.com/corelight/community-id-spec) (https://github.com/vectordotdev/vrl/pull/360)
 - updated aws vpc flow log parsing to include version 5 fields (https://github.com/vectordotdev/vrl/issues/227)
 - removed deprecated `to_timestamp` function (https://github.com/vectordotdev/vrl/pull/452)
-- changed `truncate` arguments, it now accepts a suffix string instead of a boolean (https://github.com/vectordotdev/vrl/pull/454)
```
Comment (MichaHoffmann) [inline CHANGELOG.md:27]: Formatter again, but the change seems ok; can i keep it?

File: Cargo.toml (hunk 1 above)
Comment (MichaHoffmann) [reply Cargo.toml]: Done!

File: CHANGELOG.md (hunk 2 above)
Comment (pront) [reply CHANGELOG.md:27]: Yes. No problem.

**Review by pront — CHANGES_REQUESTED** [review]
1. Please run `dd-rust-license-tool write`
2. Ensure `./scripts/checks.sh` passes
3. Please make a new PR [here](https://github.com/vectordotdev/vector/tree/master/website/cue/reference/remap/functions) to document these two new fucntions.

**Comment by MichaHoffmann** [comment]
> 1. Please run `dd-rust-license-tool write`
> 
>     2. Ensure `./scripts/checks.sh` passes
> 
>     3. Please make a new PR [here](https://github.com/vectordotdev/vector/tree/master/website/cue/reference/remap/functions) to document these two new fucntions.

Oh damn, i dont think i can get "wasm32-unknown-unknown" to work; i repaired the other checks though. Do you have an idea what to do about wasm?

**Comment by pront** [comment]
> > 1. Please run `dd-rust-license-tool write`
> >    
> >    2. Ensure `./scripts/checks.sh` passes
> >    3. Please make a new PR [here](https://github.com/vectordotdev/vector/tree/master/website/cue/reference/remap/functions) to document these two new fucntions.
> 
> Oh damn, i dont think i can get "wasm32-unknown-unknown" to work; i repaired the other checks though. Do you have an idea what to do about wasm?

You can use `#[cfg(not(target_arch = "wasm32"))]`.

**Comment by MichaHoffmann** [comment]
> `#[cfg(not(target_arch = "wasm32"))]`.

Ah! I see there is precedent for it with other functions. Thats awesome, ill fix it tomorrow, thank you!

**Comment by jszwedko** [comment]
Had we considered using https://github.com/BurntSushi/rust-snappy instead? That seems to be a native Rust implementation and so may be easier to get working under WASM.

**Comment by pront** [comment]
Indeed, https://crates.io/crates/snap is more mature.

https://crates.io/crates/snappy looks abandoned.

**Comment by MichaHoffmann** [comment]
> Indeed, https://crates.io/crates/snap is more mature.
> 
> https://crates.io/crates/snappy looks abandoned.

Converted to snap! Thanks for pointing it out!

File: src/stdlib/decode_snappy.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        );
+
+        engine.decode(text).expect("Cannot decode from Base64")
+    }
+
+    test_function![
+        decode_snappy => DecodeSnappy;
+
+        right_snappy {
+            args: func_args![value: value!(decode_base64("LKxUaGUgcXVpY2sgYnJvd24gZm94IGp1bXBzIG92ZXIgMTMgbGF6eSBkb2dzLg==").as_bytes())],
```
Comment (pront) [inline src/stdlib/decode_snappy.rs:89]: Do we need to go through base64?

For example see [tests here](https://github.com/BurntSushi/rust-snappy/blob/master/test/tests.rs#L60).

Comment (MichaHoffmann) [reply src/stdlib/decode_snappy.rs:89]: I dont know if its needed, but having it in decode and encode shows the duality nicely.

**Review by pront — APPROVED** [review]
Looks good

**Review by fuchsnj — APPROVED** [review]

**Pull request merged**